src/argilla/server/use_cases/responses/upsert_responses_in_bulk.py

```diff
@@ -0,0 +1,60 @@
+#  Copyright 2021-present, the Recognai S.L. team.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from typing import List
+
+from fastapi import Depends
+from sqlalchemy.ext.asyncio import AsyncSession
+
+from argilla.server.contexts import datasets
+from argilla.server.database import get_async_db
+from argilla.server.errors import future as errors
+from argilla.server.models import User
+from argilla.server.policies import RecordPolicyV1, authorize
+from argilla.server.schemas.v1.responses import Response, ResponseBulk, ResponseBulkError, ResponseUpsert
+from argilla.server.search_engine import SearchEngine, get_search_engine
+
+
+class UpsertResponsesInBulkUseCase:
+    def __init__(self, db: AsyncSession, search_engine: SearchEngine):
+        self.db = db
+        self.search_engine = search_engine
+
+    async def execute(self, responses: List[ResponseUpsert], user: User) -> List[ResponseBulk]:
+        responses_bulk_items = []
+
+        all_records = await datasets.get_records_by_ids(self.db, [item.record_id for item in responses])
+        non_empty_records = [r for r in all_records if r is not None]
+
+        await datasets.preload_records_relationships_before_validate(self.db, non_empty_records)
```

Comment:
In which cases the endpoint will receive responses for records of different datasets? Maybe we're complicating this endpoint logic assuming we will receive responses for records of different datasets and the endpoint could be `/api/v1/me/datasets/{dataset_id}/responses/bulk`.

Reply:
The preload query would be much lighter as we would only need to load one dataset and its questions with a simple select query.

Reply:
I see your point, but ideally, endpoints should be defined in a more REST definition approach. Defining a bulk endpoint under `/api/v1/.../responses` makes sense. We could limit response processing to one dataset on each request. Anyway, changes in this PR are improving the endpoint performance x2. So, I would say to review these improvements later.